### PR TITLE
Search UI overhaul

### PR DESCRIPTION
**Pull request by rdbende**
Done:
- Move search into a stack page in the main view instead of the popover
- Refactor code that should actually belong to the viewmodel

Remaining issue:
- Since the <kbd>space</kbd> key is a global shortcut for play/pause, it can't be inserted into the search bar

### CODE REVIEW

**Comment by rdbende** [comment]
> Remaining issue:
> -   Since the space key is a global shortcut for play/pause, it can't be inserted into the search bar


Disabling the action while the search view is open solved the issue.
Only TODO is to add Ctrl+F as a shortcut

**Comment by geigi** [comment]
The new search looks really cool! Thanks for implementing it. It seems like there are two small merge conflicts but other than that it looks good to me :)

**Review by geigi — APPROVED** [review]

**Comment by rdbende** [comment]
After the PRs related to UI are merged, I should really do #801. Resolving conflicts in the XML files is such a mess :(

@naglis could you give an approval on my conflict resolution?

**Review by naglis — COMMENTED** [review]
I did not notice any problems in the conflict resolution :)

A few things I have noticed about the search while testing locally:

1. `self._fs_monitor.get_book_online` is called twice for each book on each search (IIUC once when `SearchViewModel.authors` property is called and once for `SearchViewModel.readers` property).
2. "Hide unavailable books" setting is not respected for book results - if I have it enabled, I still get book results from storage that is offline (NOTE: AFAIR, this feature did not work at all prior to this PR). IIUC this is because in `SearchViewModel.search()` the `books` set is built directly from `self._model.books`, without checking if the books are online.

I believe both issues could be resolved e.g. if the search was further refactored by first getting the "online" books (if "Hide unavailable books" setting is enabled, all books otherwise) and then using it for both computing the set of books that match the query and building the author/reader results as well. This way `self._fs_monitor.get_book_online` would be called once per book. Not sure if this should be done in a separate PR.

File: cozy/ui/search_view.py
Comment (rdbende) [reply cozy/ui/search_view.py:80]: I loaded a huge library, and the current search method takes at most 20ms, so I don't think it's a problem. Even the listbox populator method is slower, it takes between 20-300 milliseconds to complete. Therefore I'd just leave this as-is.

File: cozy/ui/search_view.py
Comment (rdbende) [reply cozy/ui/search_view.py:84]: Actually, it doesn't even need a name, as it doesn't do logging.

File: cozy/ui/search_view.py
Comment (rdbende) [reply cozy/ui/search_view.py:80]: In contrast, the old search method was between 30-90ms, and displaying the results took around 180ms on average.

**Comment by rdbende** [comment]
Codeclimate is complaining about non-existent problems again :/

**Comment by rdbende** [comment]
> A few things I have noticed about the search while testing locally:
> 
> 1. `self._fs_monitor.get_book_online` is called twice for each book on each search (IIUC once when `SearchViewModel.authors` property is called and once for `SearchViewModel.readers` property).
> 
> 2. "Hide unavailable books" setting is not respected for book results - if I have it enabled, I still get book results from storage that is offline (NOTE: AFAIR, this feature did not work at all prior to this PR). IIUC this is because in `SearchViewModel.search()` the `books` set is built directly from `self._model.books`, without checking if the books are online.

Fixd.

File: cozy/ui/search_view.py
Comment (naglis) [reply cozy/ui/search_view.py]: > I loaded a huge library, and the current search method takes at most 20ms, so I don't think it's a problem.

When considering this I was imagining edge cases like the database being stored on a network drive and the connection dropping mid-query.

Comment (rdbende) [reply cozy/ui/search_view.py]: That could be a valid concern, but the list of books is loaded once in the model, and is only reloaded when a new storage location is added, so in this case it's not something that could happen.

File: cozy/view_model/search_view_model.py
Comment (naglis) [inline cozy/view_model/search_view_model.py:30]: `is_book_online()` check is done even when `show_offline_books` is True and we could skip the check altogether?

File: cozy/view_model/search_view_model.py
Comment (naglis) [inline cozy/view_model/search_view_model.py:25]: Do we need `_get_available_books` to return a set? I.e. can `self._model.books` contain duplicates or some other property of a set is used? If not and we could use `list[Book]`, we could simply return `self._model.books` if `show_offline_books` is True (avoid building a set) or do the list comprehension with filtering otherwise.

**Review by naglis — APPROVED** [review]

File: cozy/view_model/search_view_model.py
Comment (rdbende) [reply cozy/view_model/search_view_model.py:25]: Good point. We don't. I just copied the code from the authors getter, and saw no reason to use a list, if we're building a new iterable anyway, but it's indeed better to check `show_offline_books` beforehand, and don't build anything if it's True,

Comment (rdbende) [reply cozy/view_model/search_view_model.py]: Done

**Comment by rdbende** [comment]
@naglis, could you give a final approval, so that GH allows me to merge this?
It blocks many other things and will cause conflicts with other PRs, so it would be good to have it in `master`. I can do the rest in separate pull requests if needed.

**Review by naglis — APPROVED** [review]

**Comment by naglis** [comment]
> @naglis, could you give a final approval, so that GH allows me to merge this? It blocks many other things and will cause conflicts with other PRs, so it would be good to have it in `master`. I can do the rest in separate pull requests if needed.

@rdbende I think you need approval from someone with write/admin permissions (not sure how it is configured), as my approvals are [gray](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

**Comment by rdbende** [comment]
Huh, yeah. Never mind. Apparently this approval doesn't work if you don't have write access. I didn't know about it 😕

**Comment by rdbende** [comment]
Anyway, your reviews were useful, so thanks for that!

**Review by geigi — APPROVED** [review]

**Comment by geigi** [comment]
Finally merged! Thanks for the additional review @naglis and the fixes @rdbende!

**Pull request merged**
